File: docs/1.2/core/examples/node/controlPrograms.js
const chain = require('chain-sdk')

const client = new chain.Client()
const signer = new chain.HsmSigner()

client.mockHsm.keys.create().then(key => {
  signer.addKey(key.xpub, client.mockHsm.signerConnection)
  return key.xpub
}).then(xpub =>
  Promise.all([
    client.assets.create({
      alias: 'gold',
      rootXpubs: [xpub],
      quorum: 1,
    }),
    client.accounts.create({
      alias: 'alice',
      rootXpubs: [xpub],
      quorum: 1
    }),
    client.accounts.create({
      alias: 'bob',
      rootXpubs: [xpub],
      quorum: 1
    })
  ])
).then(() =>
  client.transactions.build(builder => {
    builder.issue({
      assetAlias: 'gold',
      amount: 100
    })
    builder.controlWithAccount({
      accountAlias: 'bob',
      assetAlias: 'gold',
      amount: 100
    })
  })
).then(
  template => signer.sign(template)
).then(
  signed => client.transactions.submit(signed)
).then(() => {
  // snippet create-receiver
  const aliceReceiverSerializedPromise = client.accounts.createReceiver({
    accountAlias: 'alice',
  }).then(aliceReceiver => {
    return JSON.stringify(aliceReceiver)
  })
  // endsnippet

  return aliceReceiverSerializedPromise
}).then(aliceReceiverSerialized => {
  // snippet build-transaction
  return client.transactions.build(builder => {
    builder.spendFromAccount({
      accountAlias: 'bob',
      assetAlias: 'gold',
      amount: 10
    })
    builder.controlWithReceiver({
      receiver: JSON.parse(aliceReceiverSerialized),
      assetAlias: 'gold',
      amount: 10
    })
  }).then(template => {
    return signer.sign(template)
  }).then(signed => {
    return client.transactions.submit(signed)
  })
  // endsnippet
}).then(() =>
  // snippet retire
  client.transactions.build(builder => {
    builder.spendFromAccount({
      accountAlias: 'alice',
      assetAlias: 'gold',
      amount: 10
    })
    builder.retire({
      assetAlias: 'gold',
      amount: 10
    })
  }).then(template => {
    return signer.sign(template)
  }).then(signed => {
    return client.transactions.submit(signed)
  })
  // endsnippet
).catch(err =>
  process.nextTick(() => { throw err })
)

File: docs/1.1/core/examples/java/ControlPrograms.java
import java.util.*;

import com.chain.api.*;
import com.chain.http.*;
import com.chain.signing.*;

class ControlPrograms {
  public static void main(String[] args) throws Exception {
    Client client = new Client();
    setup(client);

    // snippet create-receiver
    Receiver aliceReceiver = new Account.ReceiverBuilder()
      .setAccountAlias("alice")
      .create(client);
    String aliceReceiverSerialized = aliceReceiver.toJson();
    // endsnippet

    // snippet build-transaction
    Transaction.Template paymentToReceiver = new Transaction.Builder()
      .addAction(new Transaction.Action.SpendFromAccount()
        .setAccountAlias("bob")
        .setAssetAlias("gold")
        .setAmount(10)
      ).addAction(new Transaction.Action.ControlWithReceiver()
        .setReceiver(Receiver.fromJson(aliceReceiverSerialized))
        .setAssetAlias("gold")
        .setAmount(10)
      ).build(client);

    Transaction.submit(client, HsmSigner.sign(paymentToReceiver));
    // endsnippet

    // snippet retire
    Transaction.Template retirement = new Transaction.Builder()
      .addAction(new Transaction.Action.SpendFromAccount()
        .setAccountAlias("alice")
        .setAssetAlias("gold")
        .setAmount(10)
      ).addAction(new Transaction.Action.Retire()
        .setAssetAlias("gold")
        .setAmount(10)
      ).build(client);

    Transaction.submit(client, HsmSigner.sign(retirement));
    // endsnippet
  }

  public static void setup(Client client) throws Exception {
    MockHsm.Key key = MockHsm.Key.create(client);
    HsmSigner.addKey(key, MockHsm.getSignerClient(client));

    new Asset.Builder()
      .setAlias("gold")
      .addRootXpub(key.xpub)
      .setQuorum(1)
      .create(client);

    new Account.Builder()
      .setAlias("alice")
      .addRootXpub(key.xpub)
      .setQuorum(1)
      .create(client);

    new Account.Builder()
      .setAlias("bob")
      .addRootXpub(key.xpub)
      .setQuorum(1)
      .create(client);

    Transaction.submit(client, HsmSigner.sign(new Transaction.Builder()
      .addAction(new Transaction.Action.Issue()
        .setAssetAlias("gold")
        .setAmount(100)
      ).addAction(new Transaction.Action.ControlWithAccount()
        .setAccountAlias("bob")
        .setAssetAlias("gold")
        .setAmount(100)
      ).build(client)
    ));
  }
}
